### ISSUE TITLE
Creating an action dosen't save location info

### ISSUE THREAD

**Issue by Rio517** [MayOneUS/mayday-2.0-backend#80]
Hi,

I noticed that I cannot save location data when creating people with saving actions.  This is a urget need, @turino can you look into this?

I added a failing spec here: https://github.com/MayOneUS/mayday-2.0-backend/commit/8e7b8d5e6e014017168c3929cbd761a608521039

Thanks!

**Comment by turino** [comment]
Hey @Rio517,

Yeah, right now `state` isn't being passed in to `Person` at all and `update_location` won't do anything without a zip. I can write a quick and dirty fix, but I have a couple questions:

- is `state` being passed in as a 2-letter abbrev? (Presumably from dropdown)
- is it OK to overwrite existing location info if only `city` and/or `state` are passed in? Or is it important to check if old info matches new info and keep old info if more detailed?

**Comment by turino** [comment]
@Rio517 quick fix here: https://github.com/MayOneUS/mayday-2.0-backend/commit/557a27521bb96d62a8e14f01dccc57c731fd7b1d

**Comment by Rio517** [comment]
Great. THanks for the quite fix.   Later, we should probably compact the hash to be sure we're not over writing values w/ nil or empty strings.   Thanks for the fast response!

**Comment by turino** [comment]
@Rio517 I actually think the existing call to `compact` is problematic. If my address on file is
123 Main St, Springfield, MA 011111, and I update with city and zip: Bloomfield, PA, I want to be sure to not compact the hash and lose `address: nil`. I think we should only drop nil values if the new info matches existing data (in which case the update is unnecessary?)

**Comment by Rio517** [comment]
Valid point, but what if a matching value comes through?   like no city/state, but same zip?

**Comment by turino** [comment]
Exactly. Which is why `update_location` used to have a little more logic to it, I think. The whole method should probably be refactored at some point.

**Comment by turino** [comment]
you can see in the private methods there's `updated_zip?` function which is no longer used. That should probably be put back in. I'm happy to do the refactor. Not sure if we need to work out the specifics (e.g., when to keep/nullify old data) beforehand, though

**Comment by Rio517** [comment]
Yeah, i'd say this is more urgent than the person stuff.

**Comment by turino** [comment]
OK. So if state or zip doesn't match existing data I'm assuming we'll nullify old values. If zips match, I guess we should override existing city (if any)?

**Comment by turino** [comment]
@Rio517 what are the ways that location data is currently being passed in? Given the way `save_location` (on `Person`) used to work, I assume that we would get either a zip or an address, city, and zip. Now it seems like Action is passing in city and state. What is the frontend interface like for that? Does it have address, city, state, and zip, with all fields optional? Is there a real benefit to not requiring zip?

If both the new input and the existing data include zip code, the protocol is easy: nullify old values if zips don't match, keep old values if they do match (actually, even this gets a little complicated if you want to worry about zip+4). But if new input has city and state but no zip and existing has zip but no city, it's not clear how to proceed. If existing data has zip: 03431, state: 'NH', address: '123 main st', and new data has city: 'Keene', and state: 'NH', the correct response would be to recognize that Keene's zip is 03431 and therefore leave street address as is, but this requires extra work and probably isn't worth it, right? Better to just discard the address because we aren't sure?

So the flow would be something like this:

- is there is zip?
  - yes:
    - does it match existing zip?
      - yes: keep old data
      - no: discard old data
  - no:
    - do states match?
      - yes: check cities (same? keep, different? discard)
      - no: discard old data

A lot of this complexity goes away if we require zip.

BTW, I was wrong about `updated_zip?` no longer being used in `update_location`. It is being used, but right now it's ignoring all changes if zips match.

**Comment by turino** [comment]
@Rio517 I extracted a class to deal with all the logic of updating location. Check out this commit: https://github.com/MayOneUS/mayday-2.0-backend/commit/ef575c65d1e61f27f0dbcf2d64cc9d8221daa208

I need to re-do the tests before we can merge. Also, we should nail down the rules.

I'd like to stop calling `update_location` on `Location` class and instead call this from `PersonWithRemoteFields` (which will handle updating nb, instead of using the after_save callback on `Location`).

**Comment by Rio517** [comment]
Between these updates and your other flag, https://github.com/MayOneUS/mayday-2.0-backend/pull/91/commits/b453823d7c1d30e606e88d9b8207b490792f7dcf we should be good to go on this.
